Account with NotPrincipal should explicitly provide permission

vh05 opened the pull request:
This is the bucket policy with below combination
```
    
    Effect : DENY,
    Action: $OPERATION,
    NotPrincipal: $ACCOUNT

``` 
The permission on the "Action" for the $ACCOUNT mentioned in the "NotPrincipal" should be explicitly given.
    
Example: For "Action: get_object", if we want to DENY permission for * (all accounts) and we want to give explicit permission to any one or few accounts, then we can include that account as part of part of "NotPrincipal"
    
Check DFBUGS-1519 for steps to reproduce the issue
    
Fixes: https://issues.redhat.com/browse/DFBUGS-1519

shirady commented:
@vh05 please add tests in the file "test_s3_bucket_policy.js"

shirady commented on src/endpoint/s3/s3_rest.js:
```diff
@@ -287,14 +287,14 @@ async function authorize_request_policy(req) {
     if (account_identifier_id) {
         permission_by_id = await s3_bucket_policy_utils.has_bucket_policy_permission(
-            s3_policy, account_identifier_id, method, arn_path, req);
+            s3_policy, {account_identifier_name, account_identifier_id}, method, arn_path, req);
```
The account argument is a ID/name (string), not an object.
Continue with the flow, and you will see in `_is_principal_fit` the check of `principal.unwrap() === account`. You can add logs to better understand where you are (before the fix).

This check is only in NC and it checks the principal by ID (see "if (account_identifier_id) {").

vh05 replied:
@shirady any reason why we are passing id instead of name ? However the account name will be unique and we can use the name instead of id

shirady replied:
It is an additional way that we allow to add Principals, you can see in the [S3Ops.md](https://github.com/noobaa/noobaa-core/blob/master/docs/NooBaaNonContainerized/S3Ops.md#principal-field), the reason was to have it like in AWS for accounts, and the initial thought that was that t is part of solving the issue about account names - see issue #7734

vh05 replied:
So we should have account id work for NC as well rather than going for Account name ?

shirady replied:
No, we support both options. We support the account name for backward compatibility.

shirady commented:
@vh05 I would also start with testing on AWS the steps that are expected to happen before we change it our logs.

Please also add the full testing instructions that you manually did so we can understand what was done (you can add as comment if you prefer).

vh05 commented on src/endpoint/s3/s3_bucket_policy_utils.js:
```diff
@@ -174,8 +174,8 @@ function _is_principal_fit(account, statement) {
     for (const principal of _.flatten([statement_principal])) {
-        dbg.log1('bucket_policy: ', statement.Principal ? 'Principal' : 'NotPrincipal', ' fit?', principal, account);
-        if ((principal.unwrap() === '*') || (principal.unwrap() === account)) {
+        dbg.log1('bucket_policy: ', statement.Principal ? 'Principal' : 'NotPrincipal', ' fit?', principal, account.account_identifier_name);
```
I thought to pass it as object, now I am passing it as the seperate argument.

shirady replied:
It may not be necessary, and we also have a plan to pass the Principal as ARN to support future IAM - I can share the Jira with details ([link](https://issues.redhat.com/browse/MCGI-279)) if you want to understand the future steps that might be on these utils.
I would rethink on it, anyway, keep in mind that it should work for containerized and NC.

vh05 replied:
@shirady Should we continue in this direction and make NC compatible with ID or we take ARN approach ? According to [this](https://github.com/noobaa/noobaa-core/issues/7734), account name is an issue.

shirady replied:
I don't think you need to change this, just fix the issue in the link.
The issue about the names that you attached the link known and we have a solution for it - you can see the design of it here ([MCGI-281](https://issues.redhat.com/browse/MCGI-281)), but I don't think it is relevant.

nadavMiz commented:
@vh05 from what I understand this issue is not related to "deny statement using NotPrincipal". can you pleas change the PR name and description to reflect what you are actually doing here

vh05 commented:
> @vh05 from what I understand this issue is not related to "deny statement using NotPrincipal". can you pleas change the PR name and description to reflect what you are actually doing here

Sorry for the delayed response. I have updated the patch. Now the fix and the issue description matches I think

shirady commented on src/test/unit_tests/test_s3_bucket_policy.js:
What about fixing the test `should be able to use notPrincipal` it is in the scope of this PR?

vh05 replied:
`NotPrincipal` with `Allow` is not expected operation as per aws doc 

https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_notprincipal.html 

from AWS:
`NotPrincipal must be used with "Effect":"Deny". Using it with "Effect":"Allow" is not supported. `

shirady replied:
I agree, it is something that would be fixed in your suggested fix?

shirady commented on src/endpoint/s3/s3_bucket_policy_utils.js:
```diff
@@ -146,18 +146,31 @@ async function _is_object_version_fit(req, predicate, value) {
 }
 
-async function has_bucket_policy_permission(policy, account, method, arn_path, req) {
+async function has_bucket_policy_permission(policy, account, method, arn_path, req, account_identifier_type = 'id') {
```
Why did you choose to have id as default?
the id is currently only in NC.
Please add JSDoc to this function or a comment about the options of `account_identifier_type`. I think using a enum can help us.

vh05 replied:
IIUC, the id is considered to solve issues regarding account name and thats the updated one. We are giving exception to `name` parameter.

> I think using a enum can help us
I feel it is for default argument, string is fine

shirady replied:
As you can see in the [Jira ticket](https://issues.redhat.com/browse/MCGI-281), it is not solving it completely.
I didn't understand the quote; was it not clear? I meant to add enum so it will be easier to understand the options that we can use.

shirady commented on src/endpoint/s3/s3_bucket_policy_utils.js:
```diff
@@ -189,7 +201,23 @@ function _is_principal_fit(account, statement) {
     }
-    return statement.Principal ? principal_fit : !principal_fit;
+
+    if (statement.NotPrincipal) {
```
So after this change, the previous code is only relevant for `statement.Principal`?
If yes, then why did you keep this `let statement_principal = statement.Principal || statement.NotPrincipal;`?

vh05 replied:
If you see `statement_principal` is has in general considered to iterate through it irrespective of `Principal` or `NotPrincipal`. It is not required to be changes I feel

shirady commented on src/endpoint/s3/s3_bucket_policy_utils.js:
```diff
@@ -189,7 +201,23 @@ function _is_principal_fit(account, statement) {
-    return statement.Principal ? principal_fit : !principal_fit;
+
+    if (statement.NotPrincipal) {
+        // If the account is categorized as "Not Principal," we need to ALLOW operations
+        // for that account under the associated principal.
+        //
+        // To maintain backward compatibility, we also support name-based comparison for principal matching.
+        //
+        // - If `account_identifier_type` is "id" but the `account` parameter contains a name,
+        //   we return `false` to proceed/bypass to the next step, where we check for a name match.
+        //
+        // - In the next step, if `account_identifier_type` is "name" and the `account` also contains a name,
+        //   we return the actual value.
```
What if the allow is by ids and the deny is by names?
You can add it a test case.
What will happen when we support ARN, it will be more complex.

shirady replied:
The name is another option, we decided to not remove this to be backward compatible.
Anyway, it is a valid option, and in the containerized it is the only option.

shirady reviewed (commented):
Added a couple of comments and questions.

vh05 commented on src/test/unit_tests/test_s3_bucket_policy.js:
```diff
@@ -1647,6 +1647,37 @@ mocha.describe('s3_bucket_policy', function() {
         }));
     });
 
+    mocha.it.skip('should be able to use notPrincipal with Effect Deny', async function() {
```
Copy paste error

vh05 commented on src/test/unit_tests/test_s3_bucket_policy.js:
```diff
@@ -1647,6 +1647,37 @@ mocha.describe('s3_bucket_policy', function() {
+            Key: KEY,
+        });
+
+        await assert_throws_async(s3_a.putObject({
```
The account added in `NotPrincipal` is allowed on the `Action`. It is like excluding everyone other than  account added as `NotPrincipal`

shirady replied:
When you use `assert_throws_async` it is expected to throw an error.
As you mentioned the `user_a` is allowed on the action, then you probably need to run it without this function on `s3_a` and can run it on another user.

vh05 replied:
Yeah. That expression should be with user_b not user_a. My bad

nadavMiz commented on src/endpoint/s3/s3_bucket_policy_utils.js:
```diff
@@ -189,7 +201,23 @@ function _is_principal_fit(account, statement) {
-    return statement.Principal ? principal_fit : !principal_fit;
+
+    if (statement.NotPrincipal) {
+        // If the account is categorized as "Not Principal," we need to ALLOW operations
```
have you looked at https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_notaction.html?
note that there is an implicit Deny for actions that does not any statement. combination of Deny with NotPrincipal is not the same as Allow

vh05 committed:
Account with NotPrincipal should explicitly provide permission

This is the bucket policy with below combination

```

    Effect : DENY,
    Action: $OPERATION,
    NotPrincipal: $ACCOUNT

```
The permission on the "Action" for the $ACCOUNT mentioned in the
"NotPrincipal" should be explicitly given.

Example: For "Action: get_object", if we want to DENY permission
for * (all accounts) and we want to give explicit permission to
any one or few accounts, then we can include that account as
part of part of "NotPrincipal"

Check DFBUGS-1519 for steps to reproduce the issue

Fixes: https://issues.redhat.com/browse/DFBUGS-1519

Signed-off-by: Vinayakswami Hariharmath <vharihar@redhat.com>
